File: go/variable-length-quantity/variable_length_quantity.go
package variablelengthquantity

import (
	"errors"
)

const testVersion = 4

// DecodeVarint decodes a variable int stream of bytes into 32-bit integers
func DecodeVarint(encoded []byte) ([]uint32, error) {
	complete := true
	decoded := []uint32{}

	for _, b := range encoded {
		if complete {
			decoded = append(decoded, 0)
		}

		complete = (b & 0x80) == 0

		decoded[len(decoded)-1] <<= 7
		decoded[len(decoded)-1] |= uint32(b & 0x7f)
	}

	if !complete {
		return []uint32{}, errors.New("variablelengthquantity: incomplete sequence")
	}

	return decoded, nil
}

// EncodeVarint encodes 32-bit integer as a variable int stream of bytes
func EncodeVarint(decoded []uint32) []byte {
	encoded := []byte{}

	for _, u := range decoded {
		e := []byte{byte(u & 0x7f)}

		for {
			u >>= 7

			if u == 0 {
				break
			}

			e = append([]byte{byte(u&0x7f) | 0x80}, e...)
		}

		encoded = append(encoded, e...)
	}

	return encoded
}

File: lua/variable-length-quantity/variable-length-quantity.lua
local vlq = {}

function vlq.decode(encoded)
  decoded = {}
  value = 0
  incomplete = false

  for i, b in ipairs(encoded) do
    incomplete = true
    value = bit32.lshift(value, 7) + bit32.band(b, 0x7F)

    if not bit32.btest(b, 0x80) then
      incomplete = false
      decoded[#decoded+1] = value
      value = 0
    end
  end

  if incomplete then error('incomplete byte sequence') end

  return decoded
end

function vlq.encode(decoded)
  encoded = {}
  index = 1

  for i, value in ipairs(decoded) do
    b = 0

    repeat
      b = bit32.bor(b, bit32.band(value, 0x7F))
      table.insert(encoded, index, b)
      value = bit32.rshift(value, 7)
      b = 0x80
    until value == 0

    index = #encoded + 1
  end

  return encoded
end

return vlq
